### ISSUE TITLE
Make `no-cond-assign` work for ternaries too

### ISSUE THREAD

**Issue by fvictorio** [eslint/eslint#10091]
**What rule do you want to change?**

`no-cond-assign`

**Does this change cause the rule to produce more or fewer warnings?**

More warnings

**How will the change be implemented? (New option, new default behavior, etc.)?**

I guess new default?

**Please provide some example code that this change will affect:**

```js
  const str = (x = 0) ? 'zero' : 'non-zero'
```

**What does the rule currently do for this code?**

No warnings are produced

**What will the rule do after it's changed?**

Produce an "Unexpected assignment" warning for the condition of the ternary

**Comment by not-an-aardvark** [comment]
Thanks for the proposal. I think your example is actually different from other cases that `no-cond-assign` reports, because it's parsed like this:

```js
const str = (x = (0 ? 'zero' : 'non-zero'))
```

In other words, the "condition" of the ternary isn't the assignment `x = 0`, it's just the constant `0`. So I would expect this to be reported by `no-constant-condition`, not `no-cond-assign`. It appears that `no-constant-condition` is already reporting it with the current implementation.

**Comment by fvictorio** [comment]
You are right, sorry. I meant `(x = 0) ? 'zero' : 'non-zero'. I will edit the issue's description.

**Comment by platinumazure** [comment]
I'm not sure if this will really be a worthwhile change, due to the fact that replacing two `==` with one `=` in many cases will change the precedence rules for the statement (as outlined by not-an-aardvark [here](https://github.com/eslint/eslint/issues/10091#issuecomment-374039169)). So I'm not sure that a change here would even work in most cases.

It might be better to use `no-restricted-syntax` to catch both of these cases, using selectors similar to the following:
* `AssignmentExpression[right.type='AssignmentExpression'][right.right.type.expression='ConditionalExpression']`
* `ConditionalExpression[test.type='AssignmentExpression']`

**Comment by not-an-aardvark** [comment]
I think this is worth changing because it highlights a potential error -- in most cases, users probably didn't intend to use an assignment in a conditional.

**Comment by fvictorio** [comment]
> replacing two == with one = in many cases will change the precedence rules for the statement

Yes, but I think many developers surround the condition in the ternary with parentheses, so in many _other_ cases the rule will be useful.

**Comment by platinumazure** [comment]
I won't stand in the way of this. If we can get a champion for this rule (or if not-an-aardvark, VictorHom, or Aladdin-ADD champion and we get one more :+1: on the original issue comment), then we can accept the proposal.

I do think we might need to bikeshed on which of the cases I highlighted above should be supported by this rule. It would be good to agree on that before we accept the proposal.

**Comment by not-an-aardvark** [comment]
Could you clarify which cases you're referring to? My working hypothesis is that the rule should always report `ConditionalExpression` nodes where the `test` property is an `AssignmentExpression`.

**Comment by platinumazure** [comment]
@not-an-aardvark Agreed that we should definitely fix that case.

I'm more worried about `AssignmentExpression[right.type='AssignmentExpression'][right.right.type.expression='ConditionalExpression']`, something like `x = y = 0 ? true : false`. Does that really fit in with this rule? Hard to say.

**Comment by not-an-aardvark** [comment]
I don't think so, I think that should be handled by `no-constant-condition` (or `no-multi-assign` if the user has that enabled).

**Comment by platinumazure** [comment]
Okay, I can get behind `no-multi-assign` flagging the latter case. We'll fix `ConditionalExpression[test.type='AssignmentExpression']` here. I'll :+1: the opening comment.

**Comment by not-an-aardvark** [comment]
I'll champion it.

**Comment by fvictorio** [comment]
Do you folks think this could be a good first contribution? I'd like to give it a try.

**Comment by ljharb** [comment]
@fvictorio looks like https://github.com/eslint/eslint/pull/10109 already attempts to handle it
